micro-rdk/src/common/conn/viam.rs

```diff
@@ -0,0 +1,756 @@
+        if !self.storage.has_robot_credentials() {
+            self.provision().await;
+        }
+
+        // assume credentials are valid for now
+        let robot_creds = self.storage.get_robot_credentials().unwrap();
+
+        // The next step is to build the robot based on the cached config (or an empty Robot)
+        // 1) We are offline therefore viam server will not start webrtc listening (AppClient wil not be constructed)
+        //    However we are still able to connect locally (H2) and we should cache data is the data manager exists.
+        // 2) We are online, hence there is a chance the robot we created has the wrong config. By default we are going to
+        //    destroy the robot previously created. And reload everything.
+        // is_connected only tells us whether or not we are on a network
+        // Note default will use the same executor
+        let robot = self
+            .storage
+            .get_robot_configuration()
+            .map_or(LocalRobot::default(), |cfg| {
+                LocalRobot::from_cloud_config2(
+                    self.executor.clone(),
+                    robot_creds.robot_id.clone(),
+                    &cfg,
+                    &mut self.component_registry, // why do we need a mutable ref here?
+                    None,
+                )
+                .unwrap_or_default()
+            });
+        let robot = Arc::new(Mutex::new(robot));
+
+        let app_client = self.connect_to_app().await.unwrap();
+
+        let certs = app_client.get_certificates().await.unwrap();
+
+        if let HTTP2Server::HTTP2Connector(s) = &mut self.http2_server {
+            s.set_server_certificates(
+                certs.tls_certificate.into_bytes(),
+                certs.tls_private_key.into_bytes(),
+            );
+        }
+        let ip = self.network.get_ip();
+
+        let config = app_client.get_app_config(Some(ip)).await.unwrap();
+        self.rpc_host = config.0.config.unwrap().cloud.unwrap().fqdn;
+
+        let (tx, rx) = async_channel::bounded(1);
+        let inner = RobotServer::new(
+            &self.http2_server,
+            &self.webrtc_configuration,
+            self.executor.clone(),
+            robot.clone(),
+            &self.mdns,
+            self.http2_server_port,
+            rx,
+            &self.network,
+        );
+
+        self.app_client_tasks.push(Box::new(SignalingTask {
+            sender: tx,
+            rpc_host: self.rpc_host.clone(),
+        }));
+        let mut tasks: FuturesUnordered<_> = FuturesUnordered::new();
+        tasks.push(Either::Right(self.run_app_client_tasks(Some(app_client))));
+        tasks.push(Either::Left(inner.run()));
+        while let Some(_) = tasks.next().await {}
+    }
+    async fn connect_to_app(&self) -> Result<AppClient, AppClientError> {
+        //ugly hack to remove last /
+        //needs to change that but would need to update GrpcClient
```

Comment:
Tbh, the PR is pretty big anyway, might as well make this change as well?

Reply:
yep but technically non-blocker and might be a ticket for later if we wanted too